### ISSUE TITLE
Rotated multiscale zarrs are clipped/throw errors sometimes

### ISSUE THREAD

**Issue by ahnsws** [napari/napari#1959]
## 🐛 Bug

When calling `viewer.add_image` with a zarr pyramid, various `rotate` values result in the following:

| `rotate` in degs | Result |
| ----- | ----- |
| `0` | normal |
| `(0, 54.2]` | image appears clipped; e.g. `1` causes slight clipping, `54.2` causes nothing but sliver|
| `[54.3, 324.1]` | throws an error |
| `[324.2, 360)` | image appears clipped; e.g. `359` causes slight clipping, `324.2` causes nothing but sliver |
| `360` | normal

The errors vary, including:

- `RuntimeError: OpenGL got errors (Check before draw): GL_INVALID_VALUE`
- `RuntimeError: OpenGL got errors (periodic check): GL_INVALID_VALUE`
- `RuntimeError: Could not find VBO with id {X}`
- `RuntimeError: Cannot SIZE object {X} because it does not exist`

## To Reproduce

1. Create a zarr

```python
import numpy
import skimage.data as data
import zarr
from skimage.transform import pyramid_gaussian


def create_zarr(path: str, image: numpy.ndarray, chunk_size: int = 512) -> None:
    pyramid = pyramid_gaussian(image, downscale=2, max_layer=4, multichannel=True)

    store = zarr.DirectoryStore(path)
    with zarr.group(store, overwrite=True) as group:
        series = []
        for i, layer in enumerate(pyramid):
            path = "base" if i == 0 else f"L{i}"
            group.create_dataset(path, data=layer, chunks=(chunk_size, chunk_size, 3))
            series.append({"path": path})

        multiscales = [{"name": "pyramid", "datasets": series, "type": "pyramid"}]
        group.attrs["multiscales"] = multiscales


create_zarr("image.zarr", numpy.tile(data.astronaut(), reps=(2, 3, 1)))
```

2. View rotated pyramid

```python
import os
from typing import List

import dask.array as da
import napari


def get_pyramid(path: str) -> List[da.Array]:
    levels = ["base"] + sorted(
        [s for s in os.listdir(path) if (not s.startswith(".")) and (s != "base")]
    )
    return [da.from_zarr(f"{str(path)}/{level}") for level in levels]


def run():
    pyramid = get_pyramid("image.zarr")
    with napari.gui_qt():
        viewer = napari.Viewer()

        viewer.add_image(pyramid, rotate=54.3)


if __name__ == "__main__":
    run()
```

Adjust `rotate` accordingly.

## Expected behavior

Rotated pyramids to appear like non-rotated pyramids.

## Environment

```
napari: 0.4.0
Platform: Windows-10-10.0.19041-SP0
Python: 3.8.6 | packaged by conda-forge | (default, Oct 7 2020, 18:22:52) [MSC v.1916 64 bit (AMD64)]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.18.5
SciPy: 1.5.3
Dask: 2.30.0
VisPy: 0.6.5

GL version: 4.6.0 NVIDIA 452.57
MAX_TEXTURE_SIZE: 32768

Plugins:
- napari_plugin_engine: 0.1.8
- svg: 0.1.4
```

## Additional context

Full stacktrace of one of the OpenGL errors:

```
WARNING: Error drawing visual <Image at 0x13c351a5f40>
ERROR:root:Unhandled exception:
Traceback (most recent call last):
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\app\backends\_qt.py", line 825, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\util\event.py", line 473, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\scene\canvas.py", line 217, in on_draw
    self._draw_scene()
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\scene\canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\scene\canvas.py", line 304, in draw_visual
    node.draw()
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\scene\visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\visuals\visual.py", line 442, in draw
    self._program.draw(self._vshare.draw_mode,
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\visuals\shaders\program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\program.py", line 533, in draw
    canvas.context.flush_commands()
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\glir.py", line 572, in flush
    self._shared.flush(parser)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\glir.py", line 819, in parse
    self._parse(command)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\glir.py", line 787, in _parse
    ob.set_data(*args)
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\glir.py", line 1380, in set_data
    gl.check_error('periodic check')
  File "C:\Users\User\miniconda3\envs\deming\lib\site-packages\vispy\gloo\gl\__init__.py", line 207, in check_error
    raise err
RuntimeError: OpenGL got errors (periodic check): GL_INVALID_VALUE
```

Edit: a gif of `rotate=45`:

![rotated-pyramids](https://user-images.githubusercontent.com/17461258/100478776-3c616c00-30ba-11eb-88f0-b6e213976337.gif)

**Comment by sofroniewn** [comment]
Thanks for this report @wooahn, looks like we're having a bug with our affine transforms (rotation in this case) and our multiscale support. We'll look into it.

**Comment by manzt** [comment]
+1 I ran into this today trying to register two multiscale WSIs (providing an explicit `affine` matrix). For some transformations, just the weird clipping happens. But sometimes the clipping cuts a dimension of the image completely, and a `ValueError` is raised from `_vispy/vispy_image_layer.py`:

```
ValueError: Shape of individual tiles in multiscale (48727, 0, 4) cannot exceed GL_MAX_TEXTURE_SIZE. Rendering is currently in 2D mode
```

<img src="https://user-images.githubusercontent.com/24403730/111842517-f4b75800-88d5-11eb-8147-6b180835f40d.png" width="450" />

**Comment by sofroniewn** [comment]
Sorry no progress here yet. I also know this isn't something supported yet in our experimental tiled rendering. We'll try and get to this when we can. Thanks for giving it a bump though!

**Comment by manzt** [comment]
No problem, just wanted to bump. If someone else runs into this, I've found that a 0-90° rotations of the multiscale image progressively clip the image. This was an issue because one of my images needed to be rotated 90° and was clipped completely. As a workaround, I ended up applying `rot90` with dask and then applying affine matrix to the rotated source data (this removed the majority of the clipping in my case since the images were mostly aligned). Not ideal, but a workaround!

<img src="https://user-images.githubusercontent.com/24403730/112001499-67524e80-8af5-11eb-8888-6068505ed296.png" width="400" />
